src/wasm.h

```diff
@@ -273,6 +286,382 @@ class Literal {
+    assert(type == WasmType::f32);
+    return Literal(double(getf32()));
+  }
+  Literal truncateToI32() const {
+    assert(type == WasmType::i64);
+    return Literal((int32_t)i64);
+  }
+  Literal truncateToF32() const {
+    assert(type == WasmType::f64);
+    return Literal(float(getf64()));
+  }
+
+  Literal convertSToF32() const {
+    if (type == WasmType::i32) return Literal(float(i32));
+    if (type == WasmType::i64) return Literal(float(i64));
+    WASM_UNREACHABLE();
+  }
+  Literal convertUToF32() const {
+    if (type == WasmType::i32) return Literal(float(uint32_t(i32)));
+    if (type == WasmType::i64) return Literal(float(uint64_t(i64)));
+    WASM_UNREACHABLE();
+  }
+  Literal convertSToF64() const {
+    if (type == WasmType::i32) return Literal(double(i32));
+    if (type == WasmType::i64) return Literal(double(i64));
+    WASM_UNREACHABLE();
+  }
+  Literal convertUToF64() const {
+    if (type == WasmType::i32) return Literal(double(uint32_t(i32)));
+    if (type == WasmType::i64) return Literal(double(uint64_t(i64)));
+    WASM_UNREACHABLE();
+  }
+
+  Literal neg() const {
+    switch (type) {
+      case WasmType::i32: return Literal(i32 ^ 0x80000000);
+      case WasmType::i64: return Literal(int64_t(i64 ^ 0x8000000000000000ULL));
+      case WasmType::f32: return Literal(i32 ^ 0x80000000).castToF32();
+      case WasmType::f64: return Literal(int64_t(i64 ^ 0x8000000000000000ULL)).castToF64();
+      default: WASM_UNREACHABLE();
+    }
+  }
+  Literal abs() const {
+    switch (type) {
+      case WasmType::i32: return Literal(i32 & 0x7fffffff);
+      case WasmType::i64: return Literal(int64_t(i64 & 0x7fffffffffffffffULL));
+      case WasmType::f32: return Literal(i32 & 0x7fffffff).castToF32();
+      case WasmType::f64: return Literal(int64_t(i64 & 0x7fffffffffffffffULL)).castToF64();
+      default: WASM_UNREACHABLE();
+    }
+  }
+  Literal ceil() const {
+    switch (type) {
+      case WasmType::f32: return Literal(std::ceil(getf32()));
+      case WasmType::f64: return Literal(std::ceil(getf64()));
+      default: WASM_UNREACHABLE();
+    }
+  }
+  Literal floor() const {
+    switch (type) {
+      case WasmType::f32: return Literal(std::floor(getf32()));
+      case WasmType::f64: return Literal(std::floor(getf64()));
+      default: WASM_UNREACHABLE();
+    }
+  }
+  Literal trunc() const {
+    switch (type) {
+      case WasmType::f32: return Literal(std::trunc(getf32()));
+      case WasmType::f64: return Literal(std::trunc(getf64()));
+      default: WASM_UNREACHABLE();
+    }
+  }
+  Literal nearbyint() const {
+    switch (type) {
+      case WasmType::f32: return Literal(std::nearbyint(getf32()));
+      case WasmType::f64: return Literal(std::nearbyint(getf64()));
+      default: WASM_UNREACHABLE();
+    }
+  }
+  Literal sqrt() const {
+    switch (type) {
+      case WasmType::f32: return Literal(std::sqrt(getf32()));
+      case WasmType::f64: return Literal(std::sqrt(getf64()));
```

Comment:
I'm not super comfortable mapping the C++ STL behavior to wasm's, but I guess it'll do for now.